File: submarine-server/server-core/src/main/java/org/apache/submarine/server/model/database/service/ModelVersionService.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.submarine.server.model.database.service;

import org.apache.ibatis.session.SqlSession;
import org.apache.submarine.commons.utils.exception.SubmarineRuntimeException;
import org.apache.submarine.server.database.utils.MyBatisUtil;
import org.apache.submarine.server.model.database.entities.ModelVersionEntity;
import org.apache.submarine.server.model.database.mappers.ModelVersionMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

public class ModelVersionService {

  private static final Logger LOG = LoggerFactory.getLogger(ModelVersionService.class);

  public List<ModelVersionEntity> selectAllVersions(String name) throws SubmarineRuntimeException {
    LOG.info("Model Version select all versions:" + name);
    List<ModelVersionEntity> modelVersionEntities;
    try (SqlSession sqlSession = MyBatisUtil.getSqlSession()) {
      ModelVersionMapper mapper = sqlSession.getMapper(ModelVersionMapper.class);
      modelVersionEntities = mapper.selectAllVersions(name);
      sqlSession.commit();
    } catch (Exception e) {
      LOG.error(e.getMessage(), e);
      throw new SubmarineRuntimeException("Unable to get model version from database");
    }
    return modelVersionEntities;
  }

  public ModelVersionEntity select(String name, Integer version) throws SubmarineRuntimeException {
    LOG.info("Model Version select:" + name + " " + version.toString());
    ModelVersionEntity modelVersionEntity;
    try (SqlSession sqlSession = MyBatisUtil.getSqlSession()) {
      ModelVersionMapper mapper = sqlSession.getMapper(ModelVersionMapper.class);
      modelVersionEntity = mapper.select(name, version);
      sqlSession.commit();
    } catch (Exception e) {
      LOG.error(e.getMessage(), e);
      throw new SubmarineRuntimeException("Unable to get model version from database");
    }
    return modelVersionEntity;
  }
  public ModelVersionEntity selectWithTag(String name, Integer version) throws SubmarineRuntimeException {
    LOG.info("Model Version select with tag:" + name + " " + version.toString());
    ModelVersionEntity modelVersionEntity;
    try (SqlSession sqlSession = MyBatisUtil.getSqlSession()) {
      ModelVersionMapper mapper = sqlSession.getMapper(ModelVersionMapper.class);
      modelVersionEntity = mapper.selectWithTag(name, version);
      sqlSession.commit();
    } catch (Exception e) {
      LOG.error(e.getMessage(), e);
      throw new SubmarineRuntimeException("Unable to get model version from database");
    }
    return modelVersionEntity;
  }

  public void insert(ModelVersionEntity modelVersion) throws SubmarineRuntimeException {
    LOG.info("Model Version insert " + modelVersion.getName());
    try (SqlSession sqlSession = MyBatisUtil.getSqlSession()) {
      ModelVersionMapper mapper = sqlSession.getMapper(ModelVersionMapper.class);
      mapper.insert(modelVersion);
      sqlSession.commit();
    } catch (Exception e) {
      LOG.error(e.getMessage(), e);
      throw new SubmarineRuntimeException("Unable to insert model version from database");
    }
  }

  public void update(ModelVersionEntity modelVersion) throws SubmarineRuntimeException {
    LOG.info("Model Version update " + modelVersion.getName());
    try (SqlSession sqlSession = MyBatisUtil.getSqlSession()) {
      ModelVersionMapper mapper = sqlSession.getMapper(ModelVersionMapper.class);
      mapper.update(modelVersion);
      sqlSession.commit();
    } catch (Exception e) {
      LOG.error(e.getMessage(), e);
      throw new SubmarineRuntimeException("Unable to update model version from database");
    }
  }

  public void delete(String name, Integer version) throws SubmarineRuntimeException {
    LOG.info("Model Version delete name:" + name + ", version:" + version.toString());
    try (SqlSession sqlSession = MyBatisUtil.getSqlSession()) {
      ModelVersionMapper mapper = sqlSession.getMapper(ModelVersionMapper.class);
      mapper.delete(name, version);
      sqlSession.commit();
    } catch (Exception e) {
      LOG.error(e.getMessage(), e);
      throw new SubmarineRuntimeException("Unable to delete model version from database");
    }
  }
}

File: submarine-workbench/workbench-web/src/app/services/model-version.service.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { of, Observable, Subject } from 'rxjs';
import { BaseApiService } from '@submarine/services/base-api.service';
import { Rest } from '@submarine/interfaces';
import { ModelVersionInfo } from '@submarine/interfaces/model-version-info';
import { switchMap } from 'rxjs/operators';
import { ModelInfo } from '@submarine/interfaces/model-info';

@Injectable({
    providedIn: 'root',
})

export class ModelVersionService {

  private emitInfoSource = new Subject<string>();
  infoEmitted$ = this.emitInfoSource.asObservable();

  constructor(private baseApi: BaseApiService, private httpClient: HttpClient) {}

  emitInfo(id: string) {
    this.emitInfoSource.next(id);
  }

  querySpecificModel(name: string, version: string) : Observable<ModelVersionInfo> {
    const apiUrl = this.baseApi.getRestApi('/v1/model-version/' + name + '/' + version + '/');
    return this.httpClient.get<Rest<ModelVersionInfo>>(apiUrl).pipe(
      switchMap((res) => {
        if (res.success) {
          return of(res.result);
        } else {
          throw this.baseApi.createRequestError(res.message, res.code, apiUrl, 'get');
        }
      })
    );
  }

  queryModelAllVersions(name: string) : Observable<ModelVersionInfo[]> {
    const apiUrl = this.baseApi.getRestApi('/v1/model-version/' + name);
    return this.httpClient.get<Rest<ModelVersionInfo[]>>(apiUrl).pipe(
      switchMap((res) => {
        if (res.success) {
          return of(res.result);
        } else {
          throw this.baseApi.createRequestError(res.message, res.code, apiUrl, 'get');
        }
      })
    );
  }

  createModelVersion(modelVersionInfo: ModelVersionInfo, baseDir: string) : Observable<string> {
    const apiUrl = this.baseApi.getRestApi(`/v1/model-version?baseDir=${baseDir}`)
    return this.httpClient.post<Rest<any>>(apiUrl, modelVersionInfo).pipe(
      switchMap((res) => {
        if (res.success) {
          return of(res.message);
        }
        else {
          throw this.baseApi.createRequestError(res.message, res.code, apiUrl, 'post');
        }
      })
    )
  }

  deleteModelVersion(modelName: string, modelVersion: number){
    const apiUrl = this.baseApi.getRestApi(`/v1/model-version/${modelName}/${modelVersion}`);
    return this.httpClient.delete<Rest<any>>(apiUrl).pipe(
      switchMap((res) => {
        if (res.success) {
          return of(res.message);
        }
        else {
          throw this.baseApi.createRequestError(res.message, res.code, apiUrl, 'delete');
        }
      })
    )
  }

  createModelVersionTag(modelName: string, modelVersion: string, tag: string) : Observable<string> {
    const apiUrl = this.baseApi.getRestApi(`/v1/model-version/tag?name=${modelName}&version=${modelVersion}&tag=${tag}`);
    return this.httpClient.post<Rest<any>>(apiUrl, {}).pipe(
      switchMap((res) => {
        if (res.success) {
          return of(res.message);
        }
        else {
          throw this.baseApi.createRequestError(res.message, res.code, apiUrl, 'post');
        }
      })
    )
  }

  deleteModelVersionTag(modelName: string, modelVersion: string, tag: string) : Observable<string> {
    const apiUrl = this.baseApi.getRestApi(`/v1/model-version/tag?name=${modelName}&version=${modelVersion}&tag=${tag}`);
    return this.httpClient.delete<Rest<any>>(apiUrl).pipe(
      switchMap((res) => {
        if (res.success) {
          return of(res.message);
        }
        else {
          throw this.baseApi.createRequestError(res.message, res.code, apiUrl, 'delete');
        }
      })
    )
  }
}